service/core/kafka/server/src/main/scala/com/lightbend/lagom/internal/broker/kafka/TopicProducerActor.scala

```diff
@@ -52,79 +88,108 @@ private[lagom] class TaggedOffsetProducerActor[Message](
+                eventStreamSource.via(userlandFlow)
+            }
+        }
+      }
 
-    case (Some(endpoints: String), dao: OffsetDao) =>
-      // log is impacted by kafkaConfig.serviceName
-      val serviceName = kafkaConfig.serviceName.map(name => s"[$name]").getOrElse("")
-      log.debug("Kafka service {} located at URIs [{}] for producer of [{}]", serviceName, endpoints, topicId)
-      run(tag, endpoints, dao)
+      val (killSwitch, streamDone) = backoffSource
+        .viaMat(KillSwitches.single)(Keep.right)
+        .toMat(Sink.ignore)(Keep.both)
+        .run()
 
-    case (None, _) =>
-      // log is impacted by kafkaConfig.serviceName
-      kafkaConfig.serviceName match {
-        case Some(serviceName) => log.error("Unable to locate Kafka service named [{}]", serviceName)
-        case None              => log.error("Unable to locate Kafka brokers URIs")
-      }
+      shutdown = Some(killSwitch)
+      streamDone.pipeTo(self)
+    }
 
-      context.stop(self)
-  }
+    case EnsureActive(_) =>
+    // Yes, we are active
 
-  private def active: Receive = generalHandler.orElse {
     case Done =>
+      // This `Done` is materialization of the `Sink.ignore` above.
       log.info("Kafka producer stream for topic {} was completed.", topicId)
       context.stop(self)
+
+    case Status.Failure(e) =>
+      throw e
   }
 
-  private def run(tag: String, endpoints: String, dao: OffsetDao) = {
-    val readSideSource = eventStreamFactory(tag, dao.loadedOffset)
+  /**
+   * Every time we want to re/start the stream we need to locate the brokers and load the latest offset.
+   * The returned future can fail when the offset store is unavailable or times out or when the brokers list
+   * is empty or unconfigured, etc... In either case, the stream can't be built
+   */
+  private def eventualBrokersAndOffset(tagName: String): Future[(String, OffsetDao)] = {
+    // TODO: review the OffsetStore API. `prepare()` does a lot more than we want here. Ideally, prepare would be
+    // used when this actor starts and here we would only query for the latest offset.
+    val daoFuture: Future[OffsetDao] = offsetStore.prepare(s"topicProducer-$topicId", tagName)
+
+    // null or empty strings become None, otherwise Some[String]
+    def strToOpt(str: String) =
+      Option(str).filter(_.trim.nonEmpty)
+
+    // can be `None` if: not found using service locator (given serviceName)
+    // and no fallback `brokers` are setup in `config`
+    val serviceLookupFuture: Future[Option[String]] =
+      kafkaConfig.serviceName match {
+        case Some(name) =>
+          locateService(name)
+            .map { uris =>
+              strToOpt(UriUtils.hostAndPorts(uris))
+            }
+
+        case None =>
+          Future.successful(strToOpt(kafkaConfig.brokers))
+      }
+
+    val brokerList: Future[String] = serviceLookupFuture.flatMap {
```

Comment:
Why not a `map`?

Reply:
I want `brokerList` to be a `Success(brokers)` or `Failure(t)`, not throw an exception here. Wouldn't `map` let the exception pass?

Reply:
No, throwing an exception in Future's map gives you a failed future.

Reply:
Of course it does. 🤦🏼‍♂️

I was trying it with `val f1 = Future("123").map(throw new RuntimeException("asdf"))` which does throw the exception.

But the code we're discussing expands to an anonymous function:

```scala
  val f2 = Future("123").map {
    case "123" =>
      throw new RuntimeException("asdf")
    case _ =>
      throw new RuntimeException("qwer")
  }
```
so the exception is caught.